sdks/java/core/src/main/java/org/apache/beam/sdk/util/PCollectionViews.java

```diff
@@ -344,30 +430,42 @@ protected PCollectionViewBase(
      * Call this constructor to initialize the fields for which this base class provides
      * boilerplate accessors, with an auto-generated tag.
      */
-    protected PCollectionViewBase(
+    // TODO: make private as soon as runners are ported off indicator subclasses
+    protected SimplePCollectionView(
         Pipeline pipeline,
+        ViewFn<Iterable<WindowedValue<ElemT>>, ViewT> viewFn,
         WindowingStrategy<?, W> windowingStrategy,
         Coder<ElemT> valueCoder) {
-      this(pipeline, new TupleTag<Iterable<WindowedValue<ElemT>>>(), windowingStrategy, valueCoder);
+      this(
+          pipeline,
+          new TupleTag<Iterable<WindowedValue<ElemT>>>(),
+          viewFn,
+          windowingStrategy,
+          valueCoder);
     }
 
     /**
-     * For serialization only. Do not use directly. Subclasses should call from their own
-     * protected no-argument constructor.
+     * For serialization only. Do not use directly.
      */
     @SuppressWarnings("unused")  // used for serialization
-    protected PCollectionViewBase() {
+    protected SimplePCollectionView() {
       super();
     }
 
     @Override
-    public ViewT fromIterableInternal(Iterable<WindowedValue<?>> elements) {
+    public ViewFn<Iterable<WindowedValue<?>>, ViewT> getViewFn() {
       // Safe cast: it is required that the rest of the SDK maintain the invariant
       // that a PCollectionView is only provided an iterable for the elements of an
       // appropriately typed PCollection.
       @SuppressWarnings({"rawtypes", "unchecked"})
-      Iterable<WindowedValue<ElemT>> typedElements = (Iterable) elements;
-      return fromElements(typedElements);
+      ViewFn<Iterable<WindowedValue<?>>, ViewT> untypedViewFn = (ViewFn) viewFn;
+      return untypedViewFn;
+    }
+
+    @Override
+    @Deprecated
+    public ViewT fromIterableInternal(Iterable<WindowedValue<?>> elements) {
+      return getViewFn().apply(elements);
     }
 
     /**
```

Comment:
We should mark getTagInternal, getWindowingStrategyInternal and getCoderInternal as deprecated.